opensafely/notebook.py

```diff
@@ -0,0 +1,71 @@
+import json
+from pathlib import Path
+import os
+import subprocess
+import sys
+import time
+import webbrowser
+
+
+DESCRIPTION = "Run a jupyter notebook using the OpenSAFELY environment"
+
+def add_arguments(parser):
+    pass
+
+
+def docker(*args, **kwargs):
+    cmd = ("docker",) + args
+    check = kwargs.get("check", True)
+    capture = kwargs.get("capture_output", True)
+    newlines = kwargs.get("universal_newlines", True)
+    return subprocess.run(cmd, check=check, capture_output=capture, universal_newlines=newlines)
+
+
+def main():
+    container = None
+    pwd = Path(os.getcwd())
+    name = f"opensafely-notebook-{pwd.name}"
+    launch = [
+            "run",
+            # we initially detach, so we can find out the ip address.
+            "--detach",
+            "--rm",
+            "--init",
+            # we force tty now so we can re-attach stdin/out later
+            "-it",
+            f"--name={name}",
+            f"--hostname={name}",
+            "--label=opensafely",
+            "-v",
+            f"/{pwd}:/workspace",
+            "ghcr.io/opensafely-core/python",
+            "jupyter",
+            "lab",
+            "--ip",
+            "0.0.0.0",
+            # python image currently runs as root
+            "--allow-root",
+            # disable authentication, as it's only accessible via direct access
+            # to the docker container
+            "--LabApp.token=''",
+        ]
+
+    launch_ps = docker(*launch)
+    container = launch_ps.stdout.strip()
+    # let jupyter get started
+    time.sleep(1)
+
+    # figure out the url
+    inspect_ps = docker("inspect", container)
+    metadata = json.loads(inspect_ps.stdout)[0]
+    ip = metadata["NetworkSettings"]["IPAddress"]
+    url = f"http://{ip}:8888"
+
+    # open a webbrowser pointing to the docker container
+    webbrowser.open(url, new=2)
+
+    # show what the server has said so far
+    docker("logs", container, capture_output=False)
```

Comment:
Given you don't need to read stdout to get `container`, do you need to capture output from the initial command at all? That would also avoid the big silent wait you'll get if you run this without having the `python` image downloaded already.

Reply:
Huh, that's an interesting idea. I think it should work.

Reply:
So, this is trickier than I thought.

We are using --detach to also fork. If we want to do that in python, and preserve stdin/out, we need to use subprocess.Popen, as .run always waits till completion.

I had a quick try with Popen, but it gets ugly quickly, managing two programs writing to stdout  gets a very messy, tracebackes with weird formatting and doing funny things we stdint, forcing me to run `reset` in my terminal

Reply:
We could do this in a thread, ofc...

Reply:
Ah yes, I can see that's going to get tricky. I guess one way of avoiding the silent wait if the image needs to download would be to split the `run` into separate `create` and `start` calls. The `create` can be blocking and non-capturing so you get the download progress stuff. And then the `start` can be detached. Maybe all getting a bit complicated though.

If you did go down the threading route then you spin off a thread which just polls waiting for named container to come up, then gets its IP and opens a browser. And then the main thread just needs a single call to (non-detached) `docker run`.

Reply:
Great minds and all that, I'd already started down that thread route :)